File: 6.Z字形变换/6.go
package main

import (
	"fmt"
	"math"
)

func convert(s string, numRows int) string {
	if numRows <= 1 {
		return s
	}

	length := len(s)

	var res string
	numColumn := int(math.Ceil(float64(length) / 2 / float64(numRows - 1)))

	for i := 0; i < numRows ; i++ {
		for j := 0; j < numColumn ; j++ {
			if i + (numRows - 1) * 2 * j < length {
				res += string(s[i + (numRows - 1) * 2 * j])
			}

			if i != 0 && i != numRows - 1 && ((numRows - 1) * 2 * (j + 1) - i) < length {
				res += string(s[(numRows - 1) * 2 * (j + 1) - i])
			}
		}
	}

	return res
}

func main()  {
	a := convert("PAYPALISHIRING", 3)
	fmt.Println(a)
}

File: 6.Z字形变换/6.php
<?php
class Solution {

    /**
     * @param String $s
     * @param Integer $numRows
     * @return String
     */
    function convert($s, $numRows) {
        $res = '';

        if ($numRows == 1) {
            return $s;
        }

        $len = strlen($s);
        // 计算总共有多少列
        $numColumn = ceil($len / 2 / ($numRows - 1));
        for ($i = 0; $i < $numRows; $i++) {
            for ($j = 0; $j < $numColumn; $j++){
                // 找出排列规则 "｜"列排列规则是第i列第j行 格式为 i + (n - 1) * 2 * j
                $res .= $s[$i + ($numRows - 1) * 2 * $j] ?? '';

                if ($i != 0 && $i != $numRows - 1) {
                    // 找出排列规则 "/\"列排列规则是第i列第j行 格式为 (n - 1) * 2 * (j + 1) - i

                    $res .= $s[($numRows - 1) * 2 * ($j + 1) - $i] ?? '';
                }
            }
        }

        return $res;
    }
}

$a = new Solution();
$b = $a->convert('LEETCODEISHIRING', 2 );
var_dump($b);

'LDREOEIIECIHNTSG';